### PR TITLE
Added compiler support for CUDA codes (*.cu)

### PR DESCRIPTION
**Pull request by BeauJoh**
Uses nvcc and for .cu source files and nvlink to link all assembly files with a cuda source dependency.

### CODE REVIEW

File: mesonbuild/compilers/__init__.py [hunk 1]
```diff
@@ -146,6 +147,7 @@
     GnuDCompiler,
     LLVMDCompiler,
 )
+from .cuda import CudaCompiler
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/__init__.py]: __[Flake8]__

[F401] '.cuda.CudaCompiler' imported but unused

File: mesonbuild/compilers/compilers.py [hunk 2]
```diff
@@ -127,6 +129,13 @@ def is_library(fname):
                           'release': ['-O3'],
                           'minsize': ['-Os', '-g']}
 
+cuda_buildtype_args = {'plain': [],
+                       'debug': ['-O0', '-g'],
+                       'debugoptimized': ['-O1', '--debug'],
+                       'release': ['-O3', '-Otime'],
+                       'minsize': ['-O3', '-Ospace'],
+                      }
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/compilers.py]: __[Flake8]__

[E124] closing bracket does not match visual indentation

**Comment by jpakkane** [comment]
This is new functionality so it needs a test (preferably several). What does Cuda need to be installed to use it? Does anyone know if the Appveyor image has everything installed already?

**Comment by jpakkane** [comment]
Also, MRs should not have merge commits in them. Please get rid of that and use rebase in the future.

**Comment by BeauJoh** [comment]
Ok, I've added a test.

Yes, CUDA must be installed -- from [nvidia](https://developer.nvidia.com/cuda-downloads), and just involves updating the path:
`export PATH=/usr/local/cuda/bin:$PATH`

Please clarify on MRs -- what would you like my new commit to contain? When I google it, all I get is this Monty Python [skit](https://www.youtube.com/watch?v=HtyO4tmpPdk) :P 
I'm free to skype -- or google hangouts -- if you want to chat and walk me through what else you need from me.

Beau.

**Comment by jpakkane** [comment]
The problem with that is that the CI image is public and NVidia probably frowns quite a lot on redistributing their stuff...

> Please clarify on MRs -- what would you like my new commit to contain?

If you look at the commit list on this page's second tab, it has a commit labeled `Merge branch 'cuda_compiler'`. This must not be there. You need to get rid of it via Git rebasing. There is a lot of information on the net but roughly what you need to do is this:

1. Check out your branch
2. `git rebase --interactive 'HEAD~6'
3. Remove the line with the given commit, save, exit

Now you should have a branch that only has your commits. Force push it here to get the page updated.

**Comment by jpakkane** [comment]
I have rebased and updated this PR and opened #3919 with all the new stuff. Closing this one as the rest of the work will be done on that MR. Thanks.

**Pull request closed without merging**